fix(3UWMWA6W): use new ModSecurity variable

fzipi opened the pull request:
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

New set or rules that make use of the new `MULTIPART_PART_HEADERS` variable in ModSecurity.

fzipi commented on rules/REQUEST-922-MULTIPART-ATTACK.conf:
```diff
@@ -0,0 +1,118 @@
+    severity:'CRITICAL',\
+    setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}',\
+    chain"
+    SecRule TX:0 "!@within %{tx.allowed_request_content_type_charset}"
```
Noted, thanks!

airween commented on rules/REQUEST-922-MULTIPART-ATTACK.conf:
```diff
@@ -0,0 +1,115 @@
+
+# Only allow specific charsets on multipart text/plain content-type
+# Note: this is in phase:2 because these are headers that come in the body
+SecRule MULTIPART_PART_HEADERS "@rx ^content-type\s*:\s*text/plain;\s*charset=\"?(.*)\"?" \
```
Is there any reason (eg. for backports) why do we not use the `regexp-assemble` tool here?

fzipi replied:
Well, it is just _one_ regex. Not need to assemble to much there...

fzipi commented:
As soon as linting picks `secrules-parsing=0.2.5`should work properly.

fzipi commented:
For proper pipeline testing, we still need the release and dockers to be updated.

dune73 commented on rules/REQUEST-922-MULTIPART-ATTACK.conf:
```diff
@@ -0,0 +1,115 @@
+
+# This file is to address the 3UWMWA6W vulnerability.
+# It requires the latest version of ModSecurity 2.x/3.x or newer, where the variable
+# MULTIPART_  was introduced.
```
Full variable name.

fzipi replied:
Fixed.

theseion commented on rules/REQUEST-922-MULTIPART-ATTACK.conf:
```diff
@@ -0,0 +1,115 @@
+
+# Only allow specific charsets on multipart text/plain content-type
+# Note: this is in phase:2 because these are headers that come in the body
+SecRule MULTIPART_PART_HEADERS "@rx ^content-type\s*:\s*text/plain;\s*charset=\"?(.*)\"?" \
```
Note that this doesn't offer the same flexibility as 920600 will (doesn't account for other parameters, multiple media ranges).
```suggestion
SecRule MULTIPART_PART_HEADERS "@rx ^content-type\s*+:\s*+text/plain\s*+;\s*+charset\s*+=\s*+\"?(.*)\b\"?" \
```

fzipi replied:
We can use the exact same match here if that's better.

fzipi replied:
We are addinga data file here, that includes the matches from 920600.

theseion replied:
I looked at the spec https://www.rfc-editor.org/rfc/rfc2046#section-5.1 and it appears that only `Content-Type: <media-range>;<charset param>` is specified for multipart. No other parameters, as for the `Accept` header. So this is probably safe from the point of view of the RFC. However, there could be engines that use the same parser to read the `Content-Type` header. Then a parameter of `q=0.8`, for example, could be used to bypass this rule.

fzipi replied:
Changed to data file generation now.

theseion commented on rules/REQUEST-922-MULTIPART-ATTACK.conf:
```diff
@@ -0,0 +1,115 @@
+    severity:'CRITICAL',\
+    setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}',\
+    chain"
+    SecRule TX:1 "!@within %{tx.allowed_request_content_type_charset}" \
```
If allowed charsets contains `iso-8859-1` but not `iso-8859-15` and the charset is `iso-8859-15` then this rule will not match, even though it should. You probably need to add boundaries here (also above), e.g., 
```
SecRule TX:1 "!@within /%{tx.allowed_request_content_type_charset}/"
```

fzipi replied:
Maybe we can use something similar to what we do in 920480?

fzipi replied:
Updated.

dune73 commented:
I've taken a closer look and triggered all the rules. I did a fix to the bug bounty test 3, but now all 3 bug bounty tests by @azurit trigger the rules in this PR. So the PR basically does what it advertises.

I have also taken a close look at the rules themselves and thought about the risk. Namely in terms of false positives or service disruption.

This left a good impression. I think the rules are written in a rather defensive way. So they will trigger only the relative narrow attacks we have seen and I am confident they won't trigger any false positives. Namely because the MIME headers in question are rarely ever used in HTTP.

I did not look as deep as @theseion but there is a point to be made about evasions and if the rules are not too defensive. But given the rush, I think we should avoid big rewrites and different approaches. (-> This does not touch on the review of Max of course. That should be implemented.)

Additional thoughts: 
* Do we need the PL2 / 3 / 4 admin rules here, also given we want to integrate this into 920/21 in the future?
* Given the exotic nature of the attacks, we can consider moving the rules to PL2.

dune73 commented:
Just encountered "Pineapple. Pizza" in one of the tests. I totally expect this to fail sooner or later somewhere.

airween commented:
> Just encountered "Pineapple. Pizza" in one of the tests. I totally expect this to fail sooner or later somewhere.

[be](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920341.yaml#L24) [careful](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-930-APPLICATION-ATTACK-LFI/930110.yaml#L128) [what](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921180.yaml#L21) [you](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921180.yaml#L38) [wish](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921180.yaml#L72) [for](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/tests/regression/tests/REQUEST-921-PROTOCOL-ATTACK/921180.yaml#L89)...

dune73 commented:
Where is lint when you need it?

theseion commented on rules/REQUEST-922-MULTIPART-ATTACK.conf:
```diff
@@ -0,0 +1,116 @@
+
+# Only allow specific charsets when using "_charset_"
+# Note: this is in phase:2 because these are headers that come in the body
+SecRule &MULTIPART_PART_HEADERS:_charset_ "!@eq 0" \
```
Does this also need to be a `@rx` like below? Why isn't the `Content-Transfer-Encoding` header matched with `&` and `@eq`?

fzipi replied:
This is the difficult part to understand from the new MULTIPART_PART_HEADERS collection. 

The key in the collection is the `name` part of the `Content-Disposition` multipart header. So, for `Content-disposition: form-data; name="_charset_"`, you will find this in the correct place: `MULTIPART_PART_HEADERS:_charset_`.

But for `Content-Transfer-Encoding`, it is part of the headers that are keyed by a different name, the name of the `Content-disposition` part on where they are.

In the example tests there is this:
```
Content-Disposition: form-data; name="flavors"
Content-Transfer-Encoding: 8bit
Content-Type: text/plain; charset=utf-7
```

The `Content-Transfer-Encoding` here will be indexed the `MULTIPART_PART_HEADERS:flavors` variable. That's why we need to look for regular expressions in all keys, and we cannot use a particular index for that. Consequently, we cannot match with & because that might happen in any key in the MULTIPART_PART_HEADERS collection.

theseion commented on util/regexp-assemble/data/920600.data:
```diff
@@ -5,13 +5,18 @@
+##! starting sequence.
+##!
+##! If you update this one please update 921110.data also.
+
```
Remove this.

fzipi replied:
Removed.

theseion commented on util/regexp-assemble/data/922110-chain1.data:
```diff
@@ -0,0 +1 @@
+920600.data
```
Symlink support on Windows is brittle and requires "developer mode". See https://stackoverflow.com/questions/5917249/git-symbolic-links-in-windows/59761201#59761201.
I'd rather duplicate 920600.data, add a comment to the new file and wait for the include processor to be able to handle that.

fzipi replied:
Fine by me. Duplicated the file and add a comment.

dune73 commented:
There are still several comments open here:

* New comments by @theseion : https://github.com/coreruleset/coreruleset/pull/2769#pullrequestreview-1101760863
* Add remark, that this rule file will be integrated into 920 / 921 in the future
* Do we need PL2/3/4 admin rules here (since rule file is meant to disappear)
* Do we need these rules in PL1? The attacks are quite exotic. Would not it be better to apply at PL2 for this quick fix?

dune73 commented:
For the record: @53cur3M3 has settled the PL1/2 question with an argument in favor of PL1 on the dev Slack channel.

fzipi commented:
- Addressed comments from @theseion.
- Added remark about future integration in 920/921.
- Removed PL admin rules.
- Rules in PL1 was settled in Slack discussion.

dune73 commented:
Please replace this comment:

"It requires the latest version of ModSecurity 2.x/3.x or newer"

with

"It requests ModSecurity version 2.9.6 or 3.0.8 (or an updated version with backports 
of the security fixes in these versions) or a compatible engine supporting these changes."

lifeforms commented:
Also, I added some more text to that comment which was overwritten by a force-push.

I had something like:

```
# It requires ModSecurity version 2.9.6 or 3.0.8 (or an updated version with backports
# of the security fixes in these versions) or a compatible engine supporting these changes.
# If you cannot upgrade your engine, this file will cause ModSecurity to fail to start.
# In that case, you can temporarily delete this file. However, you will be missing
# protection from these rules. Therefore, we recommend upgrading your engine instead.
```

dune73 commented:
This sounds very good.

fzipi commented:
Ok, addressed final comments.

dune73 commented:
I think we're good. Please merge.

jurgenmehja commented:
Hi, should there be a "&" before the [rule name](https://github.com/coreruleset/coreruleset/pull/2769/commits/c8148f99e4a73e7bf1447fe485372cd0dfcbbaf3#diff-0a943a1263275e650dea75fd0610ee9627c5eed59cffdcb27bfbe49430228275R27)?

airween commented:
> Hi, should there be a "&" before the [rule name](https://github.com/coreruleset/coreruleset/pull/2769/commits/c8148f99e4a73e7bf1447fe485372cd0dfcbbaf3#diff-0a943a1263275e650dea75fd0610ee9627c5eed59cffdcb27bfbe49430228275R27)?

yes, `&` before a variable name means the count of the collection, or the number of occurrence. In this case, the rule counts the parts in the whole multipart body, where the name is `_charset_`.
